daemon/endpoint.go

```diff
@@ -69,18 +71,35 @@ func (d *Daemon) getEndpointList(params GetEndpointParams) []*models.Endpoint {
 	}
 
 	eps := d.endpointManager.GetEndpoints()
-	epModelsCh := make(chan *models.Endpoint, len(eps))
-
-	epModelsWg.Add(len(eps))
-	for _, ep := range eps {
-		go func(wg *sync.WaitGroup, epChan chan<- *models.Endpoint, ep *endpoint.Endpoint) {
-			if ep.HasLabels(convertedLabels) {
-				epChan <- ep.GetModel()
+	if len(eps) < maxGoroutines {
+		maxGoroutines = len(eps)
+	}
+	epsCh := make(chan *endpoint.Endpoint, maxGoroutines)
+	epModelsCh := make(chan *models.Endpoint, maxGoroutines)
+
+	epWorkersWg.Add(maxGoroutines)
+	for i := 0; i < maxGoroutines; i++ {
+		// Run goroutines to process each endpoint and the corresponding model.
+		// The obtained endpoint model is sent to the endpoint models channel from
+		// where it will be aggregated later.
+		go func(wg *sync.WaitGroup, epModelsChan chan<- *models.Endpoint, epsChan <-chan *endpoint.Endpoint) {
+			for ep := range epsChan {
+				if ep.HasLabels(convertedLabels) {
+					epModelsChan <- ep.GetModel()
+				}
 			}
 			wg.Done()
-		}(&epModelsWg, epModelsCh, ep)
+		}(&epWorkersWg, epModelsCh, epsCh)
 	}
 
+	// Send the endpoints to be aggregated a models to the endpoint channel.
+	go func(epsChan chan<- *endpoint.Endpoint, eps []*endpoint.Endpoint) {
```

Comment:
Since `epsChan` is a buffered channel, this doesn't need to be a go routine

Reply:
`epsModelChan` is also a buffered channel with size maxGoroutines which is used at line 108. If the number of endpoints is greater than 2 * maxGoroutines then both `epsCh` and `epsModelChan` will be filled and the mentioned function not being a goroutine will block and there will be a deadlock situation.

Reply:
@fristonio very good point. Now that I've noticed, both `epsCh` and `epModelsCh` don't need to have a length of `maxGoroutines`. Either `epModelsCh := make(chan *models.Endpoint, len(eps))` and we can remove this go routine or have the go routine and leave the `epModelsCh := make(chan *models.Endpoint, maxGoroutines)`. Either way it's good as it is now. :+1: